File: src/dom/Element.ts
import { Node } from './Node'
import { Text } from './Text'
import { camelCase, EMPTY_OBJ } from '../core/utils'
import { Document } from './Document'
import { CSSStyleDeclaration } from '../styles/CSSStyleDeclaration'

export class Element extends Node {
  id?
  style = new CSSStyleDeclaration(this.ownerDocument._scene, this._surface)
  // preact needs this sometimes
  attributes = []

  constructor(public ownerDocument: Document, public tagName, _surface) {
    super(ownerDocument, Node.ELEMENT_NODE, _surface)
  }

  // so the events can bubble
  // @see EventTarget
  _getTheParent() {
    return this.parentElement
  }

  _updateText() {
    // this is very ugly temporary hack just to have something working
    // we dont support mixing text & elements yet so we
    // just set the text to the concatenated result
    let content = '', len = this.childNodes.length
    for (let i = 0; i < len; i++) {
      const c = this.childNodes[i]

      if (c.nodeType === Node.TEXT_NODE) {
        content += (c as Text)._data
      }
    }

    this.style['content'] = content
  }

  setAttribute(name, value) {
    this[camelCase(name)] = value
  }

  removeAttribute(name) {
    delete this[camelCase(name)]
  }

  blur() {
    if (this.ownerDocument.activeElement !== this) {
      return
    }

    this._fire('blur')
    this.ownerDocument.activeElement = null
  }

  focus() {
    if (this.ownerDocument.activeElement === this) {
      return
    }

    if (this.ownerDocument.activeElement) {
      this.ownerDocument.activeElement.blur()
    }

    this.ownerDocument.activeElement = this
    this._fire('focus')
  }

  querySelector(selectors: string): Element | null {
    return this.querySelectorAll(selectors)[0] || null
  }

  // TODO: sizzle.js?
  querySelectorAll(selectors: string): Element[] {
    return []
  }

  // TODO: display: none
  get offsetParent() {
    return this.parentElement
  }

  get offsetLeft() {
    const [[left]] = this._bounds

    return left
  }

  get offsetTop() {
    const [[, top]] = this._bounds

    return top
  }

  get offsetWidth() {
    const [[left], [right]] = this._bounds

    return right - left
  }

  get offsetHeight() {
    const [[, top], [, bottom]] = this._bounds

    return bottom - top
  }

  // TODO
  get scrollLeft() {
    return 0
  }

  // TODO
  get scrollTop() {
    return 0
  }

  // TODO: "relative" to the viewport (excluding scrollX, scrollY)
  getBoundingClientRect() {
    // TODO: DOMRect
    const [[left, top], [bottom, right]] = this._bounds

    // TODO: spec allows negative width/height
    return { x: left, y: top, left, top, bottom, right, width: right - left, height: bottom - top }
  }

  get _bounds() {
    return this.ownerDocument._scene.getBounds(this._surface)
  }

  set textContent(v) {
    this.childNodes.forEach(c => c.remove())

    this.appendChild(this.ownerDocument.createTextNode(v))

    this._updateText()
  }
}

File: src/dom/element.zig
const std = @import("std");
const Node = @import("node.zig").Node;
const Document = @import("document.zig").Document;
const StyleDeclaration = @import("../css/style_declaration.zig").StyleDeclaration;
const Style = @import("../style.zig").Style;

pub const Element = struct {
    node: Node,
    local_name: []const u8,
    attributes: std.BufMap,
    style: StyleDeclaration,
    resolved_style: Style,

    pub fn init(document: *Document, local_name: []const u8) !Element {
        return .{
            .node = .{ .owner_document = document, .node_type = .element },
            .local_name = try document.allocator.dupe(u8, local_name),
            .attributes = std.BufMap.init(document.allocator),
            .style = StyleDeclaration.init(document.allocator),
            .resolved_style = .{},
        };
    }

    pub fn deinit(self: *Element) void {
        self.node.owner_document.allocator.free(self.local_name);
        self.attributes.deinit();
        self.style.deinit();
    }

    pub fn children(self: *Element) Element.ChildrenIterator {
        return .{ .nodes = self.node.childNodes() };
    }

    pub fn hasAttributes(self: *Element) bool {
        return self.attributes.hash_map.count() > 0;
    }

    pub fn hasAttribute(self: *Element, name: []const u8) bool {
        return self.attributes.hash_map.contains(name);
    }

    pub fn getAttribute(self: *Element, name: []const u8) ?[]const u8 {
        return self.attributes.get(name);
    }

    pub fn setAttribute(self: *Element, name: []const u8, value: []const u8) !void {
        try self.attributes.put(name, value);
        self.node.markDirty();
    }

    pub fn removeAttribute(self: *Element, name: []const u8) void {
        self.attributes.remove(name);
        self.node.markDirty();
    }

    pub const ChildrenIterator = struct {
        nodes: Node.ChildNodesIterator,

        pub fn next(self: *ChildrenIterator) ?*Element {
            while (self.nodes.next()) |node| {
                if (node.node_type == .element) return node.cast(Element);
            }
            return null;
        }
    };
};
